server/test/unit/paTest/handleMessageClientsTest.js

```diff
@@ -0,0 +1,266 @@
+    displayName: 'Client No Duplicate Phone Numbers',
+    responderPhoneNumbers: [successfulPhoneNumbers[4]],
+    fallbackPhoneNumbers: [successfulPhoneNumbers[5]],
+    heartbeatPhoneNumbers: [successfulPhoneNumbers[6]],
+    fromPhoneNumber: testFromPhoneNumber,
+  },
+  {
+    id: 'client-duplicate-and-multiple-phone-numbers',
+    displayName: 'Client Duplicate And Multiple Phone Numbers',
+    responderPhoneNumbers: [successfulPhoneNumbers[7]],
+    fallbackPhoneNumbers: [successfulPhoneNumbers[8], successfulPhoneNumbers[9]],
+    heartbeatPhoneNumbers: [successfulPhoneNumbers[7]],
+    fromPhoneNumber: testFromPhoneNumber,
+  },
+  {
+    id: 'client-bad-from-phone-number',
+    displayName: 'Client Bad From Phone Number',
+    responderPhoneNumbers: [failPhoneNumbers[0]],
+    fallbackPhoneNumbers: [],
+    heartbeatPhoneNumbers: [],
+    fromPhoneNumber: testBadFromPhoneNumber,
+  },
+]
+
+async function mockSendTwilioMessage(toPhoneNumber, fromPhoneNumber, twilioMessage) {
+  if (fromPhoneNumber === testBadFromPhoneNumber) {
+    return { status: 'error' }
+  }
+
+  // twilioHelpers.sendTwilioMessage does:
+  // helpers.log(`Sent by Twilio: ${response.sid}`)
+  // where 'response' comes from a Twilio API call.
+  helpers.log(`Sent by (mock) Twilio: ${twilioMessage}`)
+  return { status: 'queued' }
+}
+
+use(sinonChai)
+
+const sandbox = sinon.createSandbox()
+
+describe('pa.js unit tests: handleMessageClients', () => {
+  afterEach(() => {
+    sandbox.restore()
+  })
+
+  describe('for a request when Twilio is operating as expected', () => {
+    beforeEach(async () => {
+      sandbox.stub(db, 'getActiveClients').returns(activeClients)
+      sandbox.stub(twilioHelpers, 'sendTwilioMessage').callsFake(mockSendTwilioMessage)
+
+      this.res = mockResponse(sandbox)
+      await pa.handleMessageClients(
+        {
+          body: { twilioMessage: testTwilioMessage, googleIdToken: testGoogleIdToken },
+        },
+        this.res,
+      )
+    })
+
+    it('should respond with status 200', () => {
+      expect(this.res.status).to.be.calledWith(200)
+    })
+
+    it('should respond with the Twilio message as twilioMessage', () => {
+      expect(this.res.body.twilioMessage).to.equal(testTwilioMessage)
+    })
+
+    it('should respond with the expected phone numbers in successfullyMessaged', () => {
+      const toPhoneNumbers = this.res.body.successfullyMessaged.map(twilioTraceObject => {
+        return twilioTraceObject.to
+      })
+
+      // deep equal; no other phone numbers should be contacted
+      expect(toPhoneNumbers).to.eql(successfulPhoneNumbers)
+    })
+
+    it('should respond with the expected phone numbers in failedToMessage', () => {
+      const toPhoneNumbers = this.res.body.failedToMessage.map(twilioTraceObject => {
+        return twilioTraceObject.to
+      })
+
+      // deep equal; no other phone numbers should fail
+      expect(toPhoneNumbers).to.eql(failPhoneNumbers)
+    })
+
```

Comment:
I don't know if this is necessary, but you could add in a test case to ensure the helpers.log gets called for the failed Twilio message. Up to you though!

Reply:
I had considered this, and the only reason I didn't include that as a test was because creating the "twilioTraceObject"s from the client information in this file is kind of difficult / not very clean... In addition to this, checking that it was listed in the failedToMessage field felt sufficient in that it communicates which phone numbers, and respective client, were broken. If you think checking for helpers.log is important, though, I can definitely add checks for that!